client/scripts/views/components/search_bar.ts

```diff
@@ -225,295 +269,352 @@ const getBalancedContentListing = (unfilteredResults: any[], types: SearchType[]
 };
 
-const emptySearchPreview : m.Component<{ searchTerm: string }, {}> = {
+const emptySearchPreview: m.Component<{ searchTerm: string }, {}> = {
   view: (vnode) => {
     const { searchTerm } = vnode.attrs;
     const message = app.activeId()
       ? `No results in ${app.activeId()}. Search Commonwealth?`
       : 'No communities found';
     return m(ListItem, {
       class: 'no-results',
-      label: [
-        m('b', searchTerm),
-        m('span', message)
-      ],
+      label: [m('b', searchTerm), m('span', message)],
       onclick: (e) => {
         if (searchTerm.length < 4) {
           notifyError('Query must be at least 4 characters');
         }
         const params = `q=${encodeURIComponent(searchTerm.toString().trim())}`;
         m.route.set(`/search?${params}`);
-      }
+      },
     });
-  }
+  },
 };
 
-export const SearchBar : m.Component<{}, {
-  results: any[],
-  searchTerm: string,
-  errorText: string,
-  focused: boolean,
-  closeResults: Function,
-  hideResults: boolean,
-  inputTimeout: any,
-  isTyping: boolean,
-}> = {
+export const SearchBar: m.Component<
+  {},
+  {
+    results: any[];
+    searchTerm: string;
+    errorText: string;
+    focused: boolean;
+    closeResults: Function;
+    hideResults: boolean;
+    inputTimeout: any;
+    isTyping: boolean;
+    filterMenuActive: boolean;
+    setUsingFilterMenu: Function;
+    searchQuery: SearchQuery;
+  }
+> = {
   view: (vnode) => {
     if (!vnode.state.searchTerm) vnode.state.searchTerm = '';
+    if (!vnode.state.searchQuery) vnode.state.searchQuery = new SearchQuery('', { isSearchPreview: true });
+    if (vnode.state.searchQuery.searchTerm !== vnode.state.searchTerm && vnode.state.searchTerm.length > 3) {
+      vnode.state.searchQuery.searchTerm = vnode.state.searchTerm
+    }
+    const { results, searchTerm, searchQuery } = vnode.state;
+    const isMobile = window.innerWidth < 767.98;
+
+    const setFilterMenuActive = (using: boolean) => {
+      vnode.state.filterMenuActive = using
+    }
+
+    vnode.state.closeResults = () => {
+      vnode.state.hideResults = true;
+    };
 
-    const { results, searchTerm } = vnode.state;
-    const showDropdownPreview = !m.route.get().includes('/search?q=');
-    const isMobile = (window.innerWidth < 767.98);
-    const LoadingPreview = m(List, {
-      class: 'search-results-loading'
-    }, [ m(ListItem, { label: m(Spinner, { active: true }) }) ]);
-    const searchResults = (!results || results?.length === 0)
-      ? (app.searchCache[searchTerm]?.loaded)
-        ? m(List, [ m(emptySearchPreview, { searchTerm }) ])
-        : LoadingPreview
-      : vnode.state.isTyping
-        ? LoadingPreview
-        : m(List, { class: 'search-results-list' }, results);
-    vnode.state.closeResults = () => { vnode.state.hideResults = true; };
+    vnode.state.setUsingFilterMenu = using => {
+      vnode.state.filterMenuActive = using
+    }
+
+    const executeSearch = (query: SearchQuery) => {
```

Comment:
I wonder if the way executeSearch is scoped could be altered to solve two issues—(1) the partial indexing of searchHistory (2) that certain badly formed queries (too short, invalid chars), the user only sees a warning message if executeSearch runs [i.e. he hits enter/submit button]—which means the user here just gets an eternal loading message.

Some kind of simple typing timeout system, where after the user quits typing for X milliseconds, a "soft search" is executed (in contrast to a "hard" search) might be nice for users who prefer to stick whenever possible to the search preview, rather than changing pages/loading a full search results page.

Reply:
I added a warning in the case they don't complete their query. Not 100% positive I have the cleanest solution since Mithril acts funny when variables change in timeouts so I have to manually redraw. Let me know if you see a better way to do that.

Otherwise the preview searches are cached as discussed above